### PR TITLE
Phase out very verbose element_count functions

### PR DESCRIPTION
**Pull request by LoganDark**
This could have been done better.

I figured out the secret to declaring backwards-compatibility functions that are no longer in the header file. This doesn't work:

```cpp
RWKV_API extern "C" uint32_t rwkv_get_state_buffer_element_count(const struct rwkv_context * ctx) {
```

This works:

```cpp
extern "C" RWKV_API uint32_t rwkv_get_state_buffer_element_count(const struct rwkv_context * ctx) {
```

So now you know.

For the lulz, I left the python scripts using the old one for now just to show that they work. Let me know if you want those all updated in one go, though I'd prefer to just rework the python scripts completely (I have new bindings that behave a lot better wrt errors).

Anyway, this PR removes `rwkv_get_state_buffer_element_count` and `rwkv_get_logits_buffer_element_count` in favor of `rwkv_get_state_len` and `rwkv_get_logits_len` respectively. It also adds some new functions `rwkv_get_n_vocab`, `rwkv_get_n_embed`, and `rwkv_get_n_layer` for getting those parameters of the model, for people who are interested, and `rwkv_init_state`.

Specifically, `rwkv_get_n_vocab` will assist people automatically picking which tokenizer to use (50277 -> BPE, 65535 -> World), `rwkv_get_n_layer` will assist people in GPU offloading (figuring out how many layers are on/off GPU), `rwkv_get_n_embed` will assist people with poking at the RWKV hidden state, since some people want to do that for e.g. embeddings.

There is also a new `rwkv_init_state` which initializes a state to default / 0 tokens. Useful for people who can't / don't want to track how many times they have called `rwkv_eval` in order to pass null.

All in all this family of functions seems like a very useful and needed addition, and I don't think there'd be any problems with its inclusion.

### CODE REVIEW

File: rwkv.cpp [hunk 1]
```diff
@@ -1364,12 +1360,46 @@ bool rwkv_eval_sequence(const struct rwkv_context * ctx, const uint32_t * sequen
...
+
+size_t rwkv_get_n_embed(const struct rwkv_context * ctx) {
+    return (size_t) ctx->instance->model.header.n_embed;
+}
+
+size_t rwkv_get_n_layer(const struct rwkv_context * ctx) {
+    return (size_t) ctx->instance->model.header.n_layer;
+}
+
+size_t rwkv_state_len(const struct rwkv_context * ctx) {
```
Comment (saharNooby) [inline rwkv.cpp:1386]:
```suggestion
size_t rwkv_get_state_len(const struct rwkv_context * ctx) {
```
For consistency with other getters. Same for `rwkv_get_logits_len`

File: rwkv.cpp [hunk 2]
```diff
@@ -1364,12 +1360,46 @@ bool rwkv_eval_sequence(const struct rwkv_context * ctx, const uint32_t * sequen
...
+}
+
+void rwkv_init_state(const struct rwkv_context * ctx, float * state) {
+    const struct rwkv_file_header & header = ctx->instance->model.header;
+    const size_t layer_size = (size_t) header.n_embed * 5;
+    const size_t layer_zero = (size_t) header.n_embed * 4;
+    const size_t layers_size = (size_t) header.n_layer * layer_size;
+    for (size_t start = 0; start < layers_size; start += layer_size) {
+        for (size_t i = 0; i < layer_zero; state[start + i++] = 0.0F);
+        for (size_t i = layer_zero; i < layer_size; state[start + i++] = -1e30F);
```
Comment (saharNooby) [inline rwkv.cpp:1402]: Such loop looks confusing to me... I understood what it does after re-reading, but I believe code shound not require re-reading to understand it :) I suggest using more traditional loop, like was before

File: rwkv.h [hunk 3]
```diff
@@ -117,16 +117,36 @@ extern "C" {
...
+
+    // Returns the number of layers in the given model.
+    // Useful for always offloading the entire model to GPU, if desired.
+    RWKV_API size_t rwkv_get_n_layer(const struct rwkv_context * ctx);
+
+    // Returns the number of float elements in a complete state for the given model.
+    // This is the number of elements you'll need to allocate for a call to rwkv_eval, rwkv_eval_sequence, or rwkv_init_state.
+    RWKV_API size_t rwkv_state_len(const struct rwkv_context * ctx);
+
+    // Returns the number of elements in the logits output of a given model.
```
Comment (saharNooby) [inline rwkv.h:141]:
```suggestion
    // Returns the number of float elements in the logits output of a given model.
```

**Review by saharNooby — APPROVED** [review]

**Comment by saharNooby** [comment]
>Let me know if you want those all updated in one go

I'm sorta planning on updating Python part of the libary, specifically to support sequence mode; so I'm okay with Python part not being updated, since it will be updated later anyway.

>though I'd prefer to just rework the python scripts completely (I have new bindings that behave a lot better wrt errors)

If you already have good Python wrapper, then maybe it's better for you to contribute it, so we don't do duplicate work :) Do you have such plans?

**Comment by LoganDark** [comment]
> If you already have good Python wrapper, then maybe it's better for you to contribute it, so we don't do duplicate work :) Do you have such plans?

I do actually :3

**Comment by saharNooby** [comment]
>I do actually :3

Great! Then I will not touch the Python wrapper (library wrapper and model class) until you update it.

**Pull request merged**